### DISCUSSION TITLE
Stripping optional XML attributes

### DISCUSSION THREAD

**Discussion by dkx0stv** [TomWright/dasel#441] [Q&A] [answered]
Good morning,

the XML file below contains serial numbers of machines, each of which has a set of variables that I need to extract.

```
<response code="19" status="success">
  <result count="1" total-count="1">
    <entry name="123">
      <variable>
        <entry name="ipv4">
          <type>
            <ip>1.2.3.4</ip>
          </type>
        </entry>
        <entry name="gateway">
          <type>
            <ip>1.2.3.1</ip>
          </type>
        </entry>
      </variable>
    </entry>
    <entry admin="root" dirtyId="365" name="456" time="2024/10/23 19:36:36">
      <variable admin="root" dirtyId="365" time="2024/10/23 19:36:36">
        <entry admin="root" dirtyId="365" name="ipv4" time="2024/10/23 19:36:27">
          <type admin="root" dirtyId="365" time="2024/10/23 19:36:27">
            <ip admin="root" dirtyId="365" time="2024/10/23 19:36:27">5.6.7.8</ip>
          </type>
        </entry>
        <entry admin="root" dirtyId="365" name="gateway" time="2024/10/23 19:36:27">
          <type admin="root" dirtyId="365" time="2024/10/23 19:36:27">
            <ip admin="root" dirtyId="365" time="2024/10/23 19:36:27">5.6.7.1</ip>
          </type>
        </entry>
      </variable>
    </entry>
  </result>
</response>
```

The challenge that I am facing is that these variables may or may not be "dirty" (= not yet committed to the management server).
Efficiency matters, as the actual XML file that I am working with is ~6 MByte, I have thousands of machines, and each machine has roughly 20 variables.

I know that I could pipe the result of dasel into sed to strip the XML attributes:

```
for SERIAL in 123 456 ; do
  RESULT=$(dasel -f demo.xml ".response.result.entry.all().filter(equal(-name,$SERIAL)).variable.entry.all().filter(equal(-name,ipv4)).type.ip" | sed 's/<[^>]*>//g')
  echo $RESULT
done
```

Or I could run dasel twice:

```
for SERIAL in 123 456 ; do
  RESULT=$(dasel 2>/dev/null -f demo.xml ".response.result.entry.all().filter(equal(-name,$SERIAL)).variable.entry.all().filter(equal(-name,ipv4)).type.ip.#text")
  [ X$RESULT = "X" ] && { RESULT=$(dasel 2>/dev/null -f demo.xml ".response.result.entry.all().filter(equal(-name,$SERIAL)).variable.entry.all().filter(equal(-name,ipv4)).type.ip"); }
  echo $RESULT
done
```

However, I would like to know if there is an elegant, pure dasel solution that does not double my runtime.

Thanks a lot in advance
Martin

**Comment by TomWright** [comment]
Do you have an example expected output for the given XML above?

**Reply by dkx0stv** [reply]
I boiled the stuff that I'm doing in production down for this demo.
In this example, I am expecting the IPv4 addresses as output of my "for" loops (same result in both cases):

```
1.2.3.4
5.6.7.8
```

Thank you
Martin

**Reply by TomWright** [reply, marked as answer]
OK here's what I came up with:
```
$ dasel -f in.xml 'response.result.entry.all().filter(or(equal(-name,123),equal(-name,456))).variable.entry.all().filter(equal(-name,ipv4)).type.orDefault(ip.#text,ip)'
1.2.3.4
5.6.7.8
```

Two parts worth mentioning...

You can extend the query to include multiple serials like this if desired to save you a loop.
`filter(or(equal(-name,123),equal(-name,456)))`

The change that gives you the expected output here is: `orDefault(ip.#text,ip)`

If it can't find the `ip.#text` element, it will fallback and just return `ip`.

**Comment by dkx0stv** [comment]
Brilliant, that's it - "orDefault" solves my problem. Thank you so much!
Can I buy you a coffee?  😉

**Reply by TomWright** [reply]
Perfect, I'm glad I could help 🙂 
The best way to support me is via GitHub sponsors, it's always much appreciated! I'm pretty sure you can make one-off donations rather than a monthly amount
